core/primitives-core/src/types.rs

```diff
@@ -45,3 +45,70 @@ pub type ReceiptIndex = usize;
 pub type PromiseId = Vec<ReceiptIndex>;
 
 pub type ProtocolVersion = u32;
+
+/// The shard identifier. It may be a arbitrary number - it does not need to be
+/// a number in the range 0..NUM_SHARDS. The shard ids do not need to be
+/// sequential or contiguous.
+///
+/// The shard id is wrapped in a newtype to prevent the old pattern of using
+/// indices in range 0..NUM_SHARDS and casting to ShardId. Once the transition
+/// if fully complete it potentially may be simplified to a regular type alias.
+///
+/// TODO get rid of serde
+#[derive(
+    arbitrary::Arbitrary,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    serde::Serialize,
+    serde::Deserialize,
+    Hash,
+    Clone,
+    Copy,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+pub struct ShardId(u64);
+
+impl ShardId {
+    /// Create a new shard id. Please note that this function should not be used
+    /// directly. Instead the ShardId should be obtained from the shard_layout.
```

Comment:
It does seem slightly counter-intuitive to say `new` shouldn't be used, given the fact it's a trivial constructor. What are the risks?

Reply:
The risk is that someone may try to convert shard index to shard id using the ctor instead of using the mapping from shard layout. For example:
```
for shard_index in 1..num_shards {
  let shard_id = ShardId::new(shard_index); // Incorrect!
}
```

Reply:
I will add more context and some examples to the comment.